test/distributed/_composable/test_fully_shard.py

```diff
@@ -589,12 +599,57 @@ def _check_model_parity(self, m1: nn.Module, m2: nn.Module):
         """
         Checks that m1 and m2 have equivalent named_parameters.
         """
-        for (n1, p1), (n2, p2) in zip(
-            m1.named_parameters(), m2.named_parameters()
-        ):
+        for (n1, p1), (n2, p2) in zip(m1.named_parameters(), m2.named_parameters()):
             self.assertEqual(n1, n2)
             self.assertEqual(p1, p2)
 
 
+class TestFSDPOptimStateDict(FSDPTest):
+    """Composable FSDP optimizer state dict tests."""
+
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def _test_optim_state_dict_save_load(self):
+        orig_model = CompositeParamModel(device=torch.device("cuda"))
+
+        composable_model = copy.deepcopy(orig_model)
+        fully_shard(composable_model, policy=ModuleWrapPolicy({UnitModule}))
+        composable_optim = torch.optim.Adam(composable_model.parameters(), lr=1e-2)
+
+        orig_model = FSDP(orig_model)
+        orig_optim = torch.optim.Adam(orig_model.parameters(), lr=1e-2)
+
+        batch = torch.randn(2, 100, device="cuda")
+        for model, optim in (
+            (orig_model, orig_optim),
+            (composable_model, composable_optim),
+        ):
+            optim.zero_grad(set_to_none=True)
+            model(batch).sum().backward()
+            optim.step()
+
+        orig_optim_state_dict = FSDP._optim_state_dict(orig_model, orig_optim)
+        composable_optim_state_dict = FSDP._optim_state_dict(
+            composable_model, composable_optim
+        )
+
+        self.assertEqual(
+            len(orig_optim_state_dict["state"]),
+            len(composable_optim_state_dict["state"]),
+        )
+        for fqn, state in orig_optim_state_dict["state"].items():
+            self.assertEqual(state, composable_optim_state_dict["state"][fqn], fqn)
+
+        for group1, group2 in itertools.zip_longest(
+            orig_optim_state_dict["param_groups"],
+            composable_optim_state_dict["param_groups"],
+        ):
+            for key, value in group1.items():
+                self.assertEqual(value, group2[key])
```

Comment:
Is it worth adding tests for:

- non root FSDP
- DDP / replicate root
- nested FSDP + non root?

Reply:
Add an extra test for non root FSDP. Will add more tests after fixing the `all_gather_object` issue that prevent us from running tests on CI.